lib/aggregators/Aggregator.ts

```diff
@@ -0,0 +1,119 @@
+  public abstract result(): RDF.Term | undefined;
+
+  protected separator: string;
+  protected termTransformer: ITermTransformer;
+
+  public static emptyValue(): RDF.Term | undefined {
+    return undefined;
+  }
+
+  public constructor(expr: Algebra.AggregateExpression, protected sharedContext: ICompleteSharedContext) {
+    this.separator = expr.separator || ' ';
+    this.termTransformer = new TermTransformer(sharedContext.superTypeProvider);
+  }
+
+  protected termToNumericOrError(term: RDF.Term): E.NumericLiteral {
+    // TODO: Check behaviour
+    if (term.termType !== 'Literal') {
+      throw new Error(`Term with value ${term.value} has type ${term.termType} and is not a numeric literal`);
+    } else if (
+      !isSubTypeOf(term.datatype.value, TypeAlias.SPARQL_NUMERIC, this.sharedContext.superTypeProvider)
+    ) {
+      throw new Error(`Term datatype ${term.datatype.value} with value ${term.value} has type ${term.termType} and is not a numeric literal`);
+    }
+    return <E.NumericLiteral> this.termTransformer.transformLiteral(term);
+  }
+
+  protected extractValue(term: RDF.Term): { value: any; type: string } {
+    if (term.termType !== 'Literal') {
+      throw new Error(`Term with value ${term.value} has type ${term.termType} and is not a literal`);
+    }
+
+    const transformedLit = this.termTransformer.transformLiteral(term);
+    return { type: transformedLit.dataType, value: transformedLit.typedValue };
+  }
+}
+
+/**
+ * A base aggregator that can handle distinct and possibly wildcards.
+ */
+export class baseAggregator {
+  protected distinct: boolean;
+  protected variableValues: Map<string, Set<string | undefined>> = new Map();
+
+  public constructor(expr: Algebra.AggregateExpression, protected aggregatorComponent: AggregatorComponent) {
+    this.distinct = expr.distinct;
+  }
+
+  public static emptyValue(component: IAggregatorComponentClass): RDF.Term | undefined {
+    return component.emptyValue();
+  }
+
+  public result(): RDF.Term | undefined {
+    return this.aggregatorComponent.result();
+  }
+
+  protected put(bindings: RDF.Term | undefined, variable: string): void {
+    if (!this.canSkip(bindings, variable)) {
+      this.aggregatorComponent.put(bindings);
+      this.addSeen(bindings, variable);
+    }
+  }
+
+  protected canSkip(term: RDF.Term | undefined, variable: string): boolean {
+    const set = this.variableValues.get(variable);
+    return this.distinct && !!set && set.has(term ? RdfString.termToString(term) : undefined);
+  }
+
+  protected addSeen(term: RDF.Term | undefined, variable: string): void {
+    if (this.distinct) {
+      if (!this.variableValues.has(variable)) {
+        this.variableValues.set(variable, new Set());
+      }
+      this.variableValues.get(variable)!.add(term ? RdfString.termToString(term) : undefined);
+    }
+  }
+}
+
+/**
+ * A base aggregator that can handle distinct ignores wildcards.
+ */
+export class Aggregator extends baseAggregator {
```

Comment:
Why do we need a separate `baseAggregator`?
Can we just make this the `Aggregator`?

Reply:
Base Aggregator can make a distinction in variables, Aggregator can not, we could merge them into one and make variable a parameter with a default value of `''`. That might be cleaner?